### ISSUE TITLE
(Question) Sometimes ambient transaction is ignored

### ISSUE THREAD

**Issue by Powerz** [linq2db/linq2db#2676]
Hi guys,

I have a small question regarding the way Linq2Db treats ambient transactions (TransactionScope).

In the example (see below), if the first query runs before an instance of transaction scope is created, then the ambient transaction will be ignored. However, if the first query runs after the transaction scope is created, then the transaction will be respected.

This behavior looks more like a bug (otherwise, in my view, it deserves to be documented), but maybe it was a design decision. So is this a bug? (and if it is, would you guys fix it, please?)

### Steps to reproduce
Comment and uncomment the first line in the example posted below.

```c#
// if uncomment this line, the ambient transactions will be ignored => a new record will be committed to the database
// without this line, the ambient transaction will work => no new data will be committed to the database

// var user = await _db.Table<User>().FirstAsync();

using var transaction = new TransactionScope(TransactionScopeOption.Required, TransactionScopeAsyncFlowOption.Enabled);

await _db.InsertAsync(new User
                      {
                          Name = Guid.NewGuid().ToString(),
                          Email = $"{Guid.NewGuid()}@test.com"
                      });

throw new ApplicationException();
transaction.Complete();
```

### Environment details
linq2db version: 3.1.2, 3.1.6
Database Server: Azure Sql Server
Database Provider: SQL Server
Operating system: Windows, Mac OS
.NET Framework: .net core 3.1.300, 5.0

**Comment by MaceWindu** [comment]
In second case you create `TransactionScope` after connection already created. In such cases you need to enlist transaction in connection explicitly:
```cs
((DbConnection)_db.Connection).EnlistTransaction(Transaction.Current);
```

**Comment by MaceWindu** [comment]
linq2db doesn't call `EnlistTransaction` for user.

And I would say we shouldn't  do it. IMHO `TransactionScope` was bad design example from Microsoft.

@linq2db what do you think?

**Comment by viceroypenguin** [comment]
I would say that l2db should either completely ignore `TransactionScope` or respect the expected behavior (that creating a `TransactionScope` after the initial query still gets enlisted when the first query after the TS is created). Personally, I just use `DataConnection.CreateTransactionAsync()` and `.CommitTransactionAsync()`, since it's more convenient for async behavior than `TransactionScope`.

Maybe this could be the next analyzer - identifying uses of `TransactionScope` and issuing a warning that behavior may not be as expected.

**Comment by MaceWindu** [comment]
We cannot ignore `TransactionScope` as it works by itself - it hijacks transactions without asking anyone :smile: 
Problem in this specific case is that some functionality (specifically, attaching transaction scope to pre-existing connection) implemented in `TransactionScope` as explicit operation, which wether user or linq2db should perform.

**Comment by Powerz** [comment]
I agree with @viceroypenguin -- all or nothing. Otherwise, there will aways be some wiggle room left for confusion.

While in the above example it's possible to guess what is going on because the first (commented out) db call is explicit, in general it might be a bit tricky thing to figure out.

**Comment by MaceWindu** [comment]
Ok, we could try to do it.
Some notes:
- we should call `EnlistTransaction` only for providers that support it as default implementation throws exception
- there is no method to check that connection already enlisted, but at least for SqlClient `EnlistTransaction` implementation accept multiple calls and we should check if it is true for other providers

**Comment by Powerz** [comment]
Sounds good, thank you very much!
